test/TestUserAcceptance.hs

```diff
@@ -295,33 +242,46 @@ cleanUpTmuxSession sessionname = do
 
 
 -- | Run all application steps in a session defined by session name.
-tmuxSession :: String -> [ApplicationStep] -> TestTree
-tmuxSession tcname xs =
-  withResource setUp tearDown $
-  \_ -> testCaseSteps tcname $ \step -> runSteps step xs
-
-runSteps :: (String -> IO ()) -> [ApplicationStep] -> IO ()
-runSteps stepfx steps =
-    mapM_
-        (\a ->
-              do stepfx (asDescription a)
-                 out <- performStep "purebredtest" a
-                 d <- lookupEnv "DEBUG"
-                 when (isJust d) $ hPutStr stderr ("\n\n" ++ asDescription a ++ "\n\n" ++ out)
-                 ((asAssertInOutput a) out (asExpected a)))
-        steps
-
-performStep :: String -> ApplicationStep -> IO (String)
-performStep sessionname (ApplicationStep keys _ asLiteral expect _) = do
-    callProcess "tmux" $ communicateSessionArgs keys asLiteral
-    baton <- newEmptyMVar
-    out <- waitForString baton sessionname expect
-    _ <- takeMVar baton
-    pure out
+withTmuxSession :: TestName -> ((String -> IO ()) -> ReaderT Env IO ()) -> TestTree
+withTmuxSession tcname testfx =
+    withResource setUp tearDown $
+      \env -> testCaseSteps tcname $ \stepfx -> env >>= runReaderT (testfx stepfx)
+
+sendKeys :: String -> String -> ReaderT Env IO (String)
+sendKeys keys expect = do
+    sessionname <- getSessionName
+    liftIO $ do callProcess "tmux" $ communicateSessionArgs keys False
+                wait sessionname expect
+
+sendLiteralKeys :: String -> ReaderT Env IO (String)
+sendLiteralKeys keys = do
+    sessionname <- getSessionName
+    liftIO $ do callProcess "tmux" $ communicateSessionArgs keys True
+                wait sessionname keys
+
+capture :: ReaderT Env IO (String)
+capture = do
+  sessionname <- getSessionName
+  liftIO $ readProcess "tmux" ["capture-pane", "-e", "-p", "-t", sessionname] []
+
+getSessionName :: ReaderT Env IO (String)
+getSessionName = view (_3 . ask)
 
 holdOffTime :: Int
 holdOffTime = 10^6
 
+debugOutput :: String -> IO ()
+debugOutput out = do
+  d <- lookupEnv "DEBUG"
+  when (isJust d) $ hPutStr stderr ("\n\n" ++ out)
+
+wait :: String -> String -> IO (String)
```

Comment:
Currently the tests make a screen shot for each step. I'll loose that with this refactor. Although, it's rather just a matter of putting it somewhere else.

One thing which can be frustrating to debug is when an action times out, because whatever string we were waiting for to appear didn't appear (crashed, different characters, whatever). Having a dump in that instance is really handy. So, loosing the fact that we don't have screenshots for each passing step is not much of a big deal, if I can make one in an instance when the wait function times out.

The `waitForString` function currently invokes itself recursively until it times out. I wonder if I can refactor this and only wait like 5 * the hold off time, raise an assertion error and dump the screen. If something jumps into your mind I'd be all ears.